vault/identity_store_storage_util.go

```diff
@@ -0,0 +1,412 @@
+package vault
+
+import (
+	"crypto/md5"
+	"fmt"
+	"hash"
+	"strconv"
+	"strings"
+
+	"github.com/golang/snappy"
+	"github.com/hashicorp/vault/helper/jsonutil"
+	"github.com/hashicorp/vault/logical"
+)
+
+const (
+	bucketCount          = 256
+	packerConfigLocation = "packerconfig"
+)
+
+// storagePacker packs the logical storage entries into a specific number of
+// buckets by hashing its key and indexing based on the hash value. Currently
+// this supports only 256 bucket entries and hence relies on the first byte of
+// the hash value for indexing.
+type storagePacker struct {
+	config *storagePackerConfig
+	view   logical.Storage
+}
+
+// storagePackerConfig specifies the properties of the packer
+type storagePackerConfig struct {
+	Location   string    `json:"location" structs:"location" mapstructure:"location"`
+	HashFunc   hash.Hash `json:"hash_func" structs:"hash_func" mapstructure:"hash_func"`
+	ViewPrefix string    `json:"view_prefix" structs:"view_prefix" mapstructure:"view_prefix"`
+	NumBuckets int       `json:"num_buckets" structs:"num_buckets" mapstructure:"num_buckets"`
+}
+
+// configStorageEntry contains the properties of packer that needs to survive
+// reboot cycles
+type configStorageEntry struct {
+	NumBuckets int    `json:"num_buckets" structs:"num_buckets" mapstructure:"num_buckets"`
+	ViewPrefix string `json:"view_prefix" structs:"view_prefix" mapstructure:"view_prefix"`
+}
+
+// storageBucketEntry represents a bucket which holds many storage entries
+type storageBucketEntry struct {
+	Key   string                `json:"key" structs:"key" mapstructure:"key"`
+	Items []*entityStorageEntry `json:"items" structs:"items" mapstructure:"items"`
+}
+
+// View returns the storage view configured to be used by the packer
+func (s *storagePacker) View() logical.Storage {
+	return s.view
+}
+
+// Get returns a bucket entry for a given bucket entry key
+func (s *storagePacker) Get(bucketEntryKey string) (*storageBucketEntry, error) {
+	if bucketEntryKey == "" {
+		return nil, fmt.Errorf("missing bucket entry key")
+	}
+
+	// When trying to load entities during startup, scanning the view will
+	// attempt a read on the packer config file as a bucket entry; ignore it.
+	// There won't be a bucket entry with this same prefix as it is a reserved
```

Comment:
Even if we do this, we can't avoid a check similar to this. The reason is that `logical.CollectKeys` takes in a view (and not a prefix) which will step on both `/config/packer` and `/buckets/...` and pass each storage key to `Get` of storage util. `Get` will then have to distinguish the entries.